### PR TITLE
Add 4-bit weight support to depthwise conv

### PR DESCRIPTION
**Pull request by mansnils**
Unit tests are aligned with existing fully connected int4 unit tests.

### CODE REVIEW

File: Include/arm_nnfunctions.h [hunk 1]
```diff
@@ -828,6 +920,48 @@ arm_cmsis_nn_status arm_depthwise_conv_s8(const cmsis_nn_context *ctx,
...
+ * @param[in]      bias_dims       Bias tensor dimensions. Format: [C_OUT]
+ * @param[in]      bias_data       Bias data pointer. Data type: int32
+ * @param[in]      output_dims     Output tensor dimensions. Format: [N, H, W, C_OUT]
+ * @param[in, out] output_data     Output data pointer. Data type: int8
+ * @return     The function returns <code>ARM_CMSIS_NN_SUCCESS</code>
+ *
+ * @details
+ *    - Supported framework: TensorFlow Lite
+ */
+arm_cmsis_nn_status arm_depthwise_conv_s4(const cmsis_nn_context *ctx,
```
Comment (AdrianLundell) [inline Include/arm_nnfunctions.h:953]: Please fix these documentation errors

**Review by AdrianLundell — CHANGES_REQUESTED** [review]
Nice work! Let's just fix the documentation warnings before merging.

File: Include/arm_nnfunctions.h [hunk 2]
```diff
@@ -743,6 +743,54 @@ arm_cmsis_nn_status arm_depthwise_conv_wrapper_s8(const cmsis_nn_context *ctx,
...
+ * @param[in]      bias_data       Bias data pointer. Data type: int32
+ * @param[in]      output_dims     Output tensor dimensions. Format: [1, H, W, C_OUT]
+ * @param[in, out] output_data     Output data pointer. Data type: int8
+ * @return     The function returns
+ *                <code>ARM_CMSIS_NN_SUCCESS</code>   -  Successful completion.
+ *
+ * @details
+ *    - Supported framework: TensorFlow Lite
+ *    - Picks one of the the following functions
+ *        -# arm_depthwise_conv_s4()
```
Comment (felix-johnny) [inline Include/arm_nnfunctions.h:777]: Updatiing this part would quite likely get missed if we add a new function. I suggest to remove the listing of the functions that it picks from.

File: Include/arm_nnsupportfunctions.h [hunk 3]
```diff
@@ -787,6 +802,29 @@ __STATIC_FORCEINLINE const int8_t *read_and_pad_s4(const int8_t *source, int32_t
...
+ */
+__STATIC_FORCEINLINE const int8_t *read_and_pad_s4_uneven(const int8_t *source, int32_t *out1, int32_t *out2)
+{
+    int32_t inA1 = (source[0] & 0xFF) | ((source[1] & 0xFF) << 16);
+    int32_t inA2 = (source[1] & 0xFF) | ((source[2] & 0xFF) << 16);
+
+    *out1 = SXTB16_RORn(__sxtb16(inA2 << 4), 4);
+    *out2 = SXTB16_RORn(__sxtb16(inA1), 4);
+
+    return source;
```
Comment (felix-johnny) [inline Include/arm_nnsupportfunctions.h:825]: What is the intention of returning the unchanged source pointer?

File: Include/arm_nnfunctions.h (hunk 1 above)
Comment (mansnils) [reply Include/arm_nnfunctions.h:953]: Done

File: Include/arm_nnfunctions.h (hunk 2 above)
Comment (mansnils) [reply Include/arm_nnfunctions.h]: Done

File: Include/arm_nnsupportfunctions.h (hunk 3 above)
Comment (mansnils) [reply Include/arm_nnsupportfunctions.h]: Good point, changing this.. Done

**Review by AdrianLundell — APPROVED** [review]

**Pull request merged**